### ISSUE TITLE
Add sync_backref param to relationship

### ISSUE THREAD

**Issue by zzzeek** [sqlalchemy/sqlalchemy#5237]
We have two complaints that the change in #5149 prevents the creation of two bidirectional viewonly relationships using the backref() function.      So we will propose new keyword argment viewonly_backref=backref() which explicitly sets up a relationship in the other direction as it does now, *except* that it doesn't set up the mutation events.   viewonly_back_populates will also be added to suit the case where the bidrectionality is set up in separate relationships however we still want the lazy loader to be able to use the many-to-one loader optimization case as well as having the two relationships be introspectable as related.

**Comment by zzzeek** [comment]
see https://github.com/sqlalchemy/sqlalchemy/issues/5149#issuecomment-609082350 and https://github.com/sqlalchemy/sqlalchemy/issues/5229#issuecomment-610036657

**Comment by zzzeek** [comment]
Here are the steps:

1. relationship(..., viewonly=True) can feature a viewonly backref:

    relationship(..., viewonly=True, viewonly_backref=backref(...))

2. this can also work with a non-viewonly relationship, it will set up the reverse property and just not set up the backref events:

   relationship(..., viewonly_backref=backref(..))

3. viewonly_back_populates will allow for either / both sides being viewonly:

   bar = relationship(Bar, viewonly=True, viewonly_back_populates='foo')
   foo = relationship(Foo, viewonly=True, viewonly_back_populates='bar')

while this is kind of verbose, unfortunately there's not another way to do this without breaking backwards compatibility right now as back_populates still sets up the population events even if viewonly=True.

3. _reverse_property etc. gets set up as it does now with the above.  this actually has an impact on how the lazy loader works so this is actually sort of important which is part of why i think viewonly_back_populates needs to be there also (this for the case that @maparent requests).

4. viewonly props should *not* send the "backref" keyword into register_attribute_impl which is the part that sets up the write events, this is in orm/strategies.py, but *ONLY* in the cases where 1 and 2 are set up; it should NOT change backwards compat behavior with applications that are currently getting the warning, because we don't want to change any behavior yet (except for point 2 above which is a very slight change).  that is, an existing app that sets viewonly on one side and not the other, and generates the warning, should still be setting up the backref events.

All of the above must have tests, likely adding a new suite in test/orm/test_backref_mutations.py.    There need to be tests against:   o2m -> m2o, m2m -> m2m, o2o->m2o.    there only needs to be one collection type tested.    The test should likely use automated combinations which might be generated as:
```

for left, right in [
    ("o2m", "m2o"),
    ("m2m", "m2m"),
    ("o2o", "m2o")
]:
    for lw, rw in [
        ("writable", "viewonly"),
        ("viewonly", "writeable"),
        ("viewonly", "viewonly")
    ]:
        for back_populates_or_backref in [
            "viewonly_backref",
            "viewonly_back_populates"
        ]:
            print("test_%s_%s_%s_%s_%s" % (
                left, lw,
                right, rw,
                back_populates_or_backref
            ))
```

to produce:

```
test_o2m_writable_m2o_viewonly_viewonly_backref
test_o2m_writable_m2o_viewonly_viewonly_back_populates
test_o2m_viewonly_m2o_writeable_viewonly_backref
test_o2m_viewonly_m2o_writeable_viewonly_back_populates
test_o2m_viewonly_m2o_viewonly_viewonly_backref
test_o2m_viewonly_m2o_viewonly_viewonly_back_populates
test_m2m_writable_m2m_viewonly_viewonly_backref
test_m2m_writable_m2m_viewonly_viewonly_back_populates
test_m2m_viewonly_m2m_writeable_viewonly_backref
test_m2m_viewonly_m2m_writeable_viewonly_back_populates
test_m2m_viewonly_m2m_viewonly_viewonly_backref
test_m2m_viewonly_m2m_viewonly_viewonly_back_populates
test_o2o_writable_m2o_viewonly_viewonly_backref
test_o2o_writable_m2o_viewonly_viewonly_back_populates
test_o2o_viewonly_m2o_writeable_viewonly_backref
test_o2o_viewonly_m2o_writeable_viewonly_back_populates
test_o2o_viewonly_m2o_viewonly_viewonly_backref
test_o2o_viewonly_m2o_viewonly_viewonly_back_populates

```

**Comment by zzzeek** [comment]
cc @CaselIT @maparent @leorochael

**Comment by leorochael** [comment]
Hi @zzzeek, thanks for this.

I'm still confused by the need of `viewonly_back_populates`. I imagine it will require the respective `relationship` on the target model to already exist, just like `back_populates`. In that case:

 - Will it accept a `backref()` result? If yes, won't it maybe conflict with the definition of the relationship in the target model?
 - If not shouldn't `viewonly=` be defined in the respective relationship instead?
 - And how could you even "back populate" something that is `viewonly`? Or maybe I misunderstand and `viewonly` only affects the explicit side of the relationship where it's mentioned, not the possibility of another relationship to modify it indirectly.

In general I'm unclear why `viewonly_backref=` can't simply be done as `backref=backref(..., viewonly=True)` and `viewonly_backpopulates=` be done by adding adding `viewonly=True` directly to the `other` side of the relationship.

Unless of course it's a matter of preserving (potentially incongruent) backward compatibility behaviors. I might simply not have completely grokked what you explained above.

Anyway, thanks again for allowing for the convenience of specifying both sides of a relationship in a single stroke even in `viewonly` situations. I'm happy no matter which way this goes!

**Comment by zzzeek** [comment]
> Hi @zzzeek, thanks for this.
> 
> I'm still confused by the need of `viewonly_back_populates`. I imagine it will require the respective `relationship` on the target model to already exist, just like `back_populates`. In that case:
> 

there are two things that back_populates does that don't relate to copying data between the two relationships.   one is the case where one wants to look at the relationship.back_populates attribute in order to help with introspection of a model.  the code example in https://github.com/sqlalchemy/sqlalchemy/issues/5149#issuecomment-609067964 refers to this use case.     the other thing that can happen is if one is using a particular combination of eager and lazy loading, the lazy loader can in some cases make an optimization to not emit eager loads for a many-to-one relationship when it emits SQL as it knows the target objects will be locally present in the session.

>     * Will it accept a `backref()` result? If yes, won't it maybe conflict with the definition of the relationship in the target model?

there's two flags proposed here, one is "viewonly_backref" which accepts the "backref()" just the way "backref=backref()" does now.  The other is "viewonly_back_populates" which accepts a string and works like "back_populates" does now.

> 
>     * If not shouldn't `viewonly=` be defined in the respective relationship instead?

ideally however we are making an API here that will not conflict with the existing one that is still copying data into viewonly relationships.    also it's more explicit for one to say "viewonly_back_populates" rather than "back_populates", as these two flags have different behavior; one of them sets up event listeners that copy data and the other one will not.   the most critical thing is to avoid confusion and behaviors that silently switch on and off based on other observations.


> 
>     * And how could you even "back populate" something that is `viewonly`? 

right, it doesn't, the "viewonly_back_populates" can be named differently if some other name is feasible...mostly though this is to suit folks who want to still use backref().

> 
> 
> In general I'm unclear why `viewonly_backref=` can't simply be done as `backref=backref(..., viewonly=True)` and `viewonly_backpopulates=` be done by adding adding `viewonly=True` directly to the `other` side of the relationship.

because it would be a silent behavioral change to the current behavior, no longer cross-populating the collections/attributes, and also, it is better that a differently-behaving option have a different name.

> 
> Unless of course it's a matter of preserving (potentially incongruent) backward compatibility behaviors. I might simply not have completely grokked what you explained above.
> 
> Anyway, thanks again for allowing for the convenience of specifying both sides of a relationship in a single stroke even in `viewonly` situations. I'm happy no matter which way this goes!

**Comment by maparent** [comment]
Thank you very much for this as well, this is excellent news!

**Comment by leorochael** [comment]
@zzzeek. Thank you for the replies!

Just wanted to add a reminder to change the wording of the currently existing `viewonly` warnings to point to the newly introduced keywords.

**Comment by zzzeek** [comment]
yes, not having those warnings lead to more issues here would be the goal :)

**Comment by jvanasco** [comment]
If this change is made, shouldn't the tests also cover situations where an exception should be raised?

* `viewonly_backref` and `viewonly_back_populates` are both provided
* `viewonly_backref` is set up, but the reverse side does not have `viewonly=True`
* `viewonly_back_populates` is set up, but the reverse side does not have `viewonly=True`

**Comment by CaselIT** [comment]
I'm trying to understand all the possible options:

| viewonly | backref | back_p. | vo_backref | vo_back_p. | result | - |
|----------|---------|----------------|------------------|-------------------------|--------|---|
| false    | set     |                |                  |                         | ok, events added ||
| false    |         | set            |                  |                         | ok, events added  ||
| false    |         |                | set              |                         | ok, event not added | raise warning? |
| false    |         |                |                  | set                     | ok, event not added? | raise warning? |
| true     | set     |                |                  |                         | warning, event added? | if we use `backref()` with `viewonly=True`?|
| true     |         | set            |                  |                         | warning, event added? ||
| true     |         |                | set              |                         | ok, event not added ||
| true     |         |                |                  | set                     | ok, event not added? ||
| true|false| set    |                | set              |                         | error? ||
| true|false|        | set            | set              |                         | error? ||
| true|false| set    |                |                  | set                     | error? ||
| true|false|        | set            |                  | set                     | error? ||
| true|false| set    | set            |                  |                         | error? ||
| true|false|        |                | set              | set                     | error? ||

All other cases when more than one `backref | back_populates | viewonly_backref | viewonly_back_populates` would error

This seems to be overly complicated.
I think it would be simple to allow control of the synchronization events:
Add a new flag `sync_backref=None` (or another cleared name) instead of `viewonly_backref`, `viewonly_back_populates`
- `viewonly=True` implies `sync_backref=False` from 1.4. in 1.3 series it would warn about the changing behavior
- `viewonly=True` implies `sync_backref=True` would raise in 1.4.
- `viewonly=False` (and when not specified) would set `sync_backref=True`.
- `viewonly=False`, `sync_backref=False` would work, no synchronization of collections.

Not sure if the cases where `sync_backref` are not the same value in the two relations should be supported or not.

**Comment by jvanasco** [comment]
@CaselIT I have no opinion on the mechanism, but I would like to ensure Exceptions are raised if  invalid options are configured.  This is for 3 reasons:

1- It immediately educates developers on the proper usage
2- It avoids future tickets and mailing list discussions where people are trying to use the library in unintended and unsupported ways
3- It avoids future API constraints when we need to balance fixing an issue with it potentially breaking existing implementations.

Your suggestion makes sense to me, but Mike always sees more than we do.

**Comment by zzzeek** [comment]
> If this change is made, shouldn't the tests also cover situations where an exception should be raised?
> 
>     * `viewonly_backref` and `viewonly_back_populates` are both provided

this would be similar to "backref" and "back_populates" at the same time, so there should be a test for both scenarios if this is going to be added, it's not critical to start. this is raised right now for the existing options, this is in relationship:

```
        if self.back_populates:
            if backref:
                raise sa_exc.ArgumentError(
                    "backref and back_populates keyword arguments "
                    "are mutually exclusive"
                )

```

> 
>     * `viewonly_backref` is set up, but the reverse side does not have `viewonly=True`

that's actually fine.  That means you have a writeable relationship with a backref that is viewonly.   it only means that the backref is set up without the attribute events.

> 
>     * `viewonly_back_populates` is set up, but the reverse side does not have `viewonly=True`

again this is fine, back_populates is set up and the attribute events are not set up.  

viewonly_XYZ means exactly the same thing as XYZ except that the attribute events which copy data are not set up for that side.

**Comment by zzzeek** [comment]
Here is an alternate proposal:

1. we reopen #5149 and completely revert what we did.   

2. we do nothing in 1.3 for #5149.    

3. in 1.4 we implement #5149 by just not setting up an attribute event for an attribute that happens to have viewonly=True.

thoughts ?

**Comment by zzzeek** [comment]
OK I like @CaselIT 's sync_backref idea , that can solve this

**Comment by CaselIT** [comment]
I've updated it with also the case where one side is view only and one side is not.
This case can only work if the event are not propagated, so it's appropriate to have a warning in the normal side, that can be silenced by setting sync_backref to False.
Hopefully in this case the developer reads the warning, check the docs and understands why it needs to be set to false?
Setting it to sync_backref=True in this case raises or we have again the problem in #5149 where the view only relationship gets updated

**Comment by leorochael** [comment]
> OK I like @CaselIT 's sync_backref idea , that can solve this

Better rename this issue, then :wink:

**Comment by logicalicy** [comment]
If I understood this discussion (and proposal) right, with 1.4.0, will we want to use `sync_backref=False` for instances where we use `viewonly` together with `backref`? For context, I am a user of [sqlalchemy-continuum](https://github.com/kvesteri/sqlalchemy-continuum/), which has instances of `viewonly` + `backref`. Thanks 🙏

**Comment by CaselIT** [comment]
Yes, I think `viewonly=True` would imply `sync_backref=False` and will raise with `sync_backref=True`

**Comment by logicalicy** [comment]
Awesome, thanks for the update!

**Comment by CaselIT** [comment]
@zzzeek I'll start looking into this.
Where should I change the logic that manages the sync events between the relationships, so that in can be disabled according to the value of the new flag?

**Comment by zzzeek** [comment]
a simple boolean check you would pass through / detect / whatever right here:  https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/orm/strategies.py#L78

probably looks like this:
```

diff --git a/lib/sqlalchemy/orm/strategies.py b/lib/sqlalchemy/orm/strategies.py
index 2e9b2f316..45f621a96 100644
--- a/lib/sqlalchemy/orm/strategies.py
+++ b/lib/sqlalchemy/orm/strategies.py
@@ -75,7 +75,7 @@ def _register_attribute(
     # after the singleparentvalidator, mapper validator
     if useobject:
         backref = prop.back_populates
-        if backref:
+        if backref and prop.sync_backrefs:
             listen_hooks.append(
                 lambda desc, prop: attributes.backref_listeners(
                     desc, backref, uselist

```

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **master** branch:

**Add sync_backref flag in a relationship** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1913

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **rel_1_3** branch:

**Add sync_backref flag in a relationship** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1945

**Comment by sqla-tester** [comment]
**Federico Caselli** referenced this issue:

**Impty `sync_backref` flag in a viewonly relationship** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1961

**Comment by sqla-tester** [comment]
**Federico Caselli** referenced this issue:

**Impty `sync_backref` flag in a viewonly relationship** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1961

**Comment by CaselIT** [comment]
Just an update: this change was published in 1.3.17
